R/plotAbundanceDensity.R

```diff
@@ -0,0 +1,181 @@
+        rownames_to_column("Sample") 
+    # Gets coloring information if 'colour_by' is not NULL
+    if (!is.null(colour_by)) {
+        # Gets information from colData
+        colour_out <- retrieveCellInfo(object, colour_by)
+        # Mirrors back the variable name, if a partial match was used
+        colour_by <- colour_out$name
+        # Adds information to the data frame that includes all density data
+        density_data$colour_by <- colour_out$value
+    }
+    cols <- c("Sample","colour_by")[c("Sample","colour_by") %in% colnames(density_data)]
+    density_data <- density_data %>%
+        pivot_longer(cols = !cols, names_to = "Y", values_to = "X")
+    return(list(density_data = density_data, 
+                colour_by = colour_by))
+    
+}
+
+.density_plotter <- function(density_data, 
+                             xlab = NULL,
+                             ylab = NULL,
+                             colour_by = NULL,
+                             layout = "point",
+                             point_alpha = 0.6,
+                             point_shape = 124,
+                             point_size = 2,
+                             text_size = 8,
+                             add_legend = TRUE,
+                             flipped = FALSE,
+                             add_x_text = FALSE,
+                             add_border = NULL){
+    # start plotting
+    plot_out <- ggplot(density_data, aes_string(x="X", y="Y")) +
+        xlab(xlab) +
+        ylab(ylab)
+    
+    # Layout "density" or "point", "density" will be added
+    if (layout == "point"){
+        density_out <- .get_point_args(colour_by,
+                                       shape_by = NULL,
+                                       size_by = NULL,
+                                       alpha = point_alpha,
+                                       shape = point_shape,
+                                       size = point_size)
+        density_out$border <- TRUE
+        plot_out <- plot_out +
+            do.call(geom_point, density_out$args)
+    }
+    
+    # adjust point colours
+    if(!is.null(colour_by)){
+        if(density_out$border){
+            # resolve the colour for the line colours
+            plot_out <- .resolve_plot_colours(plot_out,
+                                              density_data$colour_by,
+                                              colour_by,
+                                              fill = FALSE)
+        }
+        # resolve the color for fill
+        plot_out <- .resolve_plot_colours(plot_out,
+                                          density_data$colour_by,
+                                          colour_by,
+                                          fill = TRUE)
+    }
+    plot_out <- plot_out +
+        theme_classic()
+    # add legend
+    plot_out <- .add_legend(plot_out, add_legend)
+    # flip
+    plot_out <- .flip_plot(plot_out, flipped, add_x_text)
+    return(plot_out)
+}
```

Comment:
please add a newline at the end of the file

Reply:
Added